ls1intum/Artemis
# `Plagiarism checks`: Add possibility for continuous daily plagiarism checks

### Checklist
#### General
#### Server
#### Client

### Motivation and Context


Adds daily automated checks for onboarded exercises. The rest of the flow is unchanged.

This feature introduces continuous plagiarism control (cpc). Plagiarism checks for exercises with the cpc enables will be executed automatically every night. Implemented logic uses existing plagiarism mechanisms, so the core algorithm is untouched and the results of checks started by cpc are presented in the interface familiar to existing users. This feature doesn't open plagiarism cases automatically, but they can be opened manually just like with manual plagiarism checks.

This is the 1st milestone of continuous plagiarism control (cpc) implementation. More about it can be discovered in the thesis proposal: https://confluence.ase.in.tum.de/display/ArTEMiS/MA+Jakub+Riegel?preview=/157431627/166332701/riegel_proposal.pdf

### Description

Changes:
- automated daily plagiarism checks:
  - checks are triggered once a day at night
  - checks execute sequentially  
  - subsequent submissions are checked automatically the next night
  - when cpc is enabled instructors can trigger plagiarism checks again manually
- server-stored plagiarism configuration:
  - configuration fields added to the exercise form (creation & edit)
  - existing exercises will have configurations with default values created on the fly
  - manual plagiarism checks use server stored configuration

Above imposed some technical changes:
- new boolean flag for enabling cpc in the exercise model
- new `ContinuousPlagiarismService` - manages cpc
- new `PlagiarismChecksService` - executes plagiarism checks logic for manual and automatic trigger
- new entity for storing plagiarism checks configuration with 1:1 relationship with exercise 
- many changes in integration test not directly relevant to plagiarism in order to adapt to the new model and flow

### Steps for Testing


Prerequisites:
- 1 Instructor
- 2 Students
- 1 Course
- for convenience use the `automated-plagiarism-checks-5mins` [branch](https://github.com/ls1intum/Artemis/tree/automated-plagiarism-checks-5-mins) which triggers checks every 5 minutes

1. Create programming, text or modelling exercise with continuous plagiarism control enabled (use _Enabled_ checkbox on exercise creation form)
2. Create two similar submission for the exercise
3. Wait 6 minutes
4. Open _Check plagiarism_ page in the exercise manager and verify that plagiarism was detected
5. Change one of submissions to something completely different
6. Wait 6 minutes
7. Open _Check plagiarism_ page in the exercise manager and verify that plagiarism submissions disappeared 
8. Change one of submissions to plagiarism again
9. Immediately after (before next automated checks are triggered) click on _Check plagiarism now_ button on the _Check plagiarism_ page
10. Verify that manual trigger detected plagiarism

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests


### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ContinuousPlagiarismControlService.java | 92% | ✅                           |
| PlagiarismChecksService.java | 100% | ✅                           |
| programming-exercise-update.component.ts | 80% | ✅              |
| modeling-exercise-update.component.ts | 84% | ✅              |
| text-exercise-update.component.ts | 85% | ✅              |

### Screenshots
##### Exercise form
New fields (for programming, text and modelling):
<img width="1507" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/518d804d-a18e-4f03-bad5-a6fe564e59e3">

##### Plagiarism page
The button for checking plagiarism has new copy with _now_ word. It better communicates to the user that this will be on-demand check, independent from continuous plagiarism control.

Initial view:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/a7c2df9e-3733-4e65-b70d-8e60b5b79a5a">

View with all buttons visible:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/d813cdc4-7104-497c-93c8-8bbd9177c9ea">
